ugvc/pipelines/correct_genotypes_by_imputation.py

```diff
@@ -0,0 +1,477 @@
+    pl_u_sum = np.sum(pl_unphred[1:])
+    pl_f = pl_unphred * f_gt_ar
+    pl_f_sum = np.sum(pl_f[1:])
+    pl_f_n = np.insert(pl_u_sum / pl_f_sum * pl_f[1:], 0, pl_unphred[0])
+    phred_pl_f_n = phred(pl_f_n)
+    # New genotype is determined by the lowest PL
+    sorted_ids = np.argsort(phred_pl_f_n)
+    current_gt_id = (gt_ar[:, 0] == current_gt[0]) & (gt_ar[:, 1] == current_gt[1])
+    pl_new_gt = phred_pl_f_n[current_gt_id].tolist()[0]
+    # If the lowest PL is the same as the PL assigned to the current genotype (a tie),then do
+    # not change the GT (the ties are caused by round-offs in the original PL)
+    if pl_new_gt == np.amin(phred_pl_f_n):
+        new_gt = current_gt
+    else:
+        new_gt = tuple(gt_ar[sorted_ids, :][0].tolist())
+    new_pl_ar = np.rint(phred_pl_f_n - min(phred_pl_f_n)).astype(int)
+    new_gq = new_pl_ar[sorted_ids][1] - new_pl_ar[sorted_ids][0]
+
+    return tuple(new_pl_ar.tolist()), new_gq.tolist(), new_gt
+
+
+def add_imputation_to_vcf(
+    beagle_anno_vcf: str,
+    out_vcf: str,
+    epsilon: float,
+    region: tuple[str, int, int] = None,
+    add_counting_category: bool = False,
+) -> dict:
+    """
+    Takes a vcf that is annotated with the Beagle results and outputs a new vcf with changes to the genotype, PLs, etc.
+    Input:
+    beagle_anno_vcf - Input vcf with Beagle annotations
+    out_vcf - Name of output vcf
+    epsilon - The weight of the imputation in determining the new PL. A number between 0 and 1.
+    region - Process the vcf only in the specified region. A tuple (chrom,start,end) (optional)
+    add_counting_category - Add the category of the variant, as used in the counting. For debugging purposes
+    """
+
+    counters = defaultdict(lambda: {"pass": 0, "has_non_ref_imp": 0, "imp_has_different_gt": 0, "changed_gt": 0})
+
+    logger.info("Modifying PLs and writing to %s", out_vcf)
+    with VariantFile(beagle_anno_vcf) as in_vcf_obj:
+        # Modify header in order to save previous GT and PL
+        gt_format = in_vcf_obj.header.formats["GT"]
+        in_vcf_obj.header.formats.add(
+            id="GT0",
+            number=gt_format.number,
+            type=gt_format.type,
+            description=gt_format.description + " (DeepVariant output)",
+        )
+
+        gq_format = in_vcf_obj.header.formats["GQ"]
+        in_vcf_obj.header.formats.add(
+            id="GQ0",
+            number=gq_format.number,
+            type=gq_format.type,
+            description=gq_format.description + " (DeepVariant output)",
+        )
+
+        pl_format = in_vcf_obj.header.formats["PL"]
+        in_vcf_obj.header.formats.add(
+            id="PL0",
+            number=pl_format.number,
+            type=pl_format.type,
+            description=pl_format.description + " (DeepVariant output)",
+        )
+        if add_counting_category:
+            in_vcf_obj.header.info.add(id="CAT", type="String", number=1, description="Category, as counted")

```

Comment:
Is this a new info field?
what does it represent?

Reply:
It is mainly for debugging purposes. I am counting variants that were (or not) affected by imputation, and here I add the category of the count directly to the vcf.